File: lua/algorithms/quick_sort.lua
-- Quick Sort - Divide and conquer sorting with median-of-three pivot
--
-- Time Complexity: O(n log n) average, O(n²) worst case
-- Space Complexity: O(log n)
-- Stable: No

local quick_sort = {}

local function medianOfThree(arr, left, right)
    local mid = math.floor(left + (right - left) / 2)

    if arr[left] > arr[mid] then
        arr[left], arr[mid] = arr[mid], arr[left]
    end
    if arr[left] > arr[right] then
        arr[left], arr[right] = arr[right], arr[left]
    end
    if arr[mid] > arr[right] then
        arr[mid], arr[right] = arr[right], arr[mid]
    end

    arr[mid], arr[right - 1] = arr[right - 1], arr[mid]
    return right - 1
end

local function partition(arr, left, right)
    local pivotIndex = medianOfThree(arr, left, right)
    local pivot = arr[pivotIndex]

    local i = left
    local j = pivotIndex - 1

    while true do
        while arr[i] < pivot do
            i = i + 1
        end
        while j > left and arr[j] > pivot do
            j = j - 1
        end

        if i >= j then
            break
        end

        arr[i], arr[j] = arr[j], arr[i]
        i = i + 1
        j = j - 1
    end

    arr[i], arr[pivotIndex] = arr[pivotIndex], arr[i]
    return i
end

local function partitionSimple(arr, left, right)
    local pivot = arr[right]
    local i = left - 1

    for j = left, right - 1 do
        if arr[j] <= pivot then
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]
        end
    end

    arr[i + 1], arr[right] = arr[right], arr[i + 1]
    return i + 1
end

local function quickSortRange(arr, left, right)
    if right - left < 10 then
        -- Use insertion sort for small arrays
        for i = left + 1, right do
            local key = arr[i]
            local j = i - 1
            while j >= left and arr[j] > key do
                arr[j + 1] = arr[j]
                j = j - 1
            end
            arr[j + 1] = key
        end
    elseif left < right then
        local pivotIndex = partition(arr, left, right)
        quickSortRange(arr, left, pivotIndex - 1)
        quickSortRange(arr, pivotIndex + 1, right)
    end
end

function quick_sort.sort(arr)
    if #arr > 1 then
        quickSortRange(arr, 1, #arr)
    end
    return arr
end

local function quickSortSimpleRange(arr, left, right)
    if left < right then
        local pivotIndex = partitionSimple(arr, left, right)
        quickSortSimpleRange(arr, left, pivotIndex - 1)
        quickSortSimpleRange(arr, pivotIndex + 1, right)
    end
end

function quick_sort.sortSimple(arr)
    if #arr > 1 then
        quickSortSimpleRange(arr, 1, #arr)
    end
    return arr
end

function quick_sort.sorted(arr)
    local result = {}
    for i = 1, #arr do
        result[i] = arr[i]
    end
    return quick_sort.sort(result)
end

local function quickSelectRange(arr, left, right, k)
    if left == right then
        return arr[left]
    end

    local pivotIndex = partitionSimple(arr, left, right)

    if k == pivotIndex then
        return arr[k]
    elseif k < pivotIndex then
        return quickSelectRange(arr, left, pivotIndex - 1, k)
    else
        return quickSelectRange(arr, pivotIndex + 1, right, k)
    end
end

function quick_sort.quickSelect(arr, k)
    if k < 1 or k > #arr then
        error("k out of bounds")
    end
    return quickSelectRange(arr, 1, #arr, k)
end

function quick_sort.findMedian(arr)
    return quick_sort.quickSelect(arr, math.floor((#arr + 1) / 2))
end

return quick_sort

File: julia/src/algorithms/quick_sort.jl
"""
Quick Sort - Divide and conquer sorting with median-of-three pivot

Time Complexity: O(n log n) average, O(n²) worst case
Space Complexity: O(log n)
Stable: No
"""

function _median_of_three!(arr::AbstractVector{T}, left::Int, right::Int) where T
    mid = left + (right - left) ÷ 2

    # Sort left, mid, right
    if arr[left] > arr[mid]
        arr[left], arr[mid] = arr[mid], arr[left]
    end
    if arr[left] > arr[right]
        arr[left], arr[right] = arr[right], arr[left]
    end
    if arr[mid] > arr[right]
        arr[mid], arr[right] = arr[right], arr[mid]
    end

    # Move median to right-1
    arr[mid], arr[right-1] = arr[right-1], arr[mid]
    right - 1
end

function _partition!(arr::AbstractVector{T}, left::Int, right::Int) where T
    pivot_index = _median_of_three!(arr, left, right)
    pivot = arr[pivot_index]

    i = left
    j = pivot_index - 1

    while true
        while arr[i] < pivot
            i += 1
        end
        while j > left && arr[j] > pivot
            j -= 1
        end

        if i >= j
            break
        end

        arr[i], arr[j] = arr[j], arr[i]
        i += 1
        j -= 1
    end

    arr[i], arr[pivot_index] = arr[pivot_index], arr[i]
    i
end

function _partition_simple!(arr::AbstractVector{T}, left::Int, right::Int) where T
    pivot = arr[right]
    i = left - 1

    for j in left:(right-1)
        if arr[j] <= pivot
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
        end
    end

    arr[i + 1], arr[right] = arr[right], arr[i + 1]
    i + 1
end

function _quick_sort_range!(arr::AbstractVector{T}, left::Int, right::Int) where T
    if right - left < 10
        # Use insertion sort for small arrays
        for i in (left + 1):right
            key = arr[i]
            j = i - 1
            while j >= left && arr[j] > key
                arr[j + 1] = arr[j]
                j -= 1
            end
            arr[j + 1] = key
        end
    elseif left < right
        pivot_index = _partition!(arr, left, right)
        _quick_sort_range!(arr, left, pivot_index - 1)
        _quick_sort_range!(arr, pivot_index + 1, right)
    end
end

"""
    quick_sort!(arr::AbstractVector{T}) where T

Sort array in-place using quick sort with median-of-three pivot.
"""
function quick_sort!(arr::AbstractVector{T}) where T
    if length(arr) > 1
        _quick_sort_range!(arr, 1, length(arr))
    end
    arr
end

function _quick_sort_simple_range!(arr::AbstractVector{T}, left::Int, right::Int) where T
    if left < right
        pivot_index = _partition_simple!(arr, left, right)
        _quick_sort_simple_range!(arr, left, pivot_index - 1)
        _quick_sort_simple_range!(arr, pivot_index + 1, right)
    end
end

"""
    quick_sort_simple!(arr::AbstractVector{T}) where T

Sort array using simple quick sort (rightmost pivot).
"""
function quick_sort_simple!(arr::AbstractVector{T}) where T
    if length(arr) > 1
        _quick_sort_simple_range!(arr, 1, length(arr))
    end
    arr
end

"""
    quick_sorted(arr::AbstractVector{T}) where T

Return a sorted copy using quick sort.
"""
function quick_sorted(arr::AbstractVector{T}) where T
    result = copy(arr)
    quick_sort!(result)
    result
end

function _quick_select!(arr::AbstractVector{T}, left::Int, right::Int, k::Int) where T
    if left == right
        return arr[left]
    end

    pivot_index = _partition_simple!(arr, left, right)

    if k == pivot_index
        return arr[k]
    elseif k < pivot_index
        return _quick_select!(arr, left, pivot_index - 1, k)
    else
        return _quick_select!(arr, pivot_index + 1, right, k)
    end
end

"""
    quick_select!(arr::AbstractVector{T}, k::Int) where T

Find the k-th smallest element (1-indexed).
Modifies the array.
"""
function quick_select!(arr::AbstractVector{T}, k::Int) where T
    if k < 1 || k > length(arr)
        throw(BoundsError(arr, k))
    end
    _quick_select!(arr, 1, length(arr), k)
end

"""
    find_median!(arr::AbstractVector{T}) where T

Find median of array (modifies array).
"""
function find_median!(arr::AbstractVector{T}) where T
    quick_select!(arr, (length(arr) + 1) ÷ 2)
end
